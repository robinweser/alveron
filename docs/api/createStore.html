
<!DOCTYPE HTML>
<html lang="en" >
    <head>
        <title>4.1. createStore Â· Alveron</title>
        <meta charset="UTF-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.1.0">
        <meta name="author" content="Robin Frischmann <robin@rofrischmann.de>">
        
        
    
    <link rel="stylesheet" href="../../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-prism/prism.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-advanced-emoji/emoji-website.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="Provider.html" />
    
    
    <link rel="prev" href="../API.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../../">
            
                <a href="../../">
            
                    
                    Read Me
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../Introduction.html">
            
                <a href="../Introduction.html">
            
                    
                    1. Introduction
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.2.1" data-path="../introduction/Motivation.html">
            
                <a href="../introduction/Motivation.html">
            
                    
                    1.1. Motivation
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.2" data-path="../introduction/How.html">
            
                <a href="../introduction/How.html">
            
                    
                    1.2. How It Works
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.3" data-path="../introduction/Design.html">
            
                <a href="../introduction/Design.html">
            
                    
                    1.3. Design Principles
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.4" data-path="../introduction/Caveats.html">
            
                <a href="../introduction/Caveats.html">
            
                    
                    1.4. Caveats
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.5" data-path="../introduction/Comparison.html">
            
                <a href="../introduction/Comparison.html">
            
                    
                    1.5. Comparison
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.6" data-path="../introduction/Ecosystem.html">
            
                <a href="../introduction/Ecosystem.html">
            
                    
                    1.6. Ecosystem
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.7" data-path="../introduction/Examples.html">
            
                <a href="../introduction/Examples.html">
            
                    
                    1.7. Examples
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../Concepts.html">
            
                <a href="../Concepts.html">
            
                    
                    2. Concepts
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../concepts/Model.html">
            
                <a href="../concepts/Model.html">
            
                    
                    2.1. Model
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../concepts/Actions.html">
            
                <a href="../concepts/Actions.html">
            
                    
                    2.2. Actions
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../concepts/Effects.html">
            
                <a href="../concepts/Effects.html">
            
                    
                    2.3. Effects
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.4" data-path="../concepts/Middleware.html">
            
                <a href="../concepts/Middleware.html">
            
                    
                    2.4. Middleware
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../Advanced.html">
            
                <a href="../Advanced.html">
            
                    
                    3. Advanced
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../advanced/Performance.html">
            
                <a href="../advanced/Performance.html">
            
                    
                    3.1. Performance
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../advanced/UsingHooks.html">
            
                <a href="../advanced/UsingHooks.html">
            
                    
                    3.2. Using Hooks
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="../API.html">
            
                <a href="../API.html">
            
                    
                    4. API Reference
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter active" data-level="1.5.1" data-path="createStore.html">
            
                <a href="createStore.html">
            
                    
                    4.1. createStore
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="Provider.html">
            
                <a href="Provider.html">
            
                    
                    4.2. Provider
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.3" data-path="Consumer.html">
            
                <a href="Consumer.html">
            
                    
                    4.3. Consumer
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.4" data-path="Wrapper.html">
            
                <a href="Wrapper.html">
            
                    
                    4.4. Wrapper
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.5" data-path="useAlveron.html">
            
                <a href="useAlveron.html">
            
                    
                    4.5. useAlveron
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href="../.." >4.1. createStore</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="createstore">createStore</h1>
<p>It creates a new store instance that provides both <a href="Provider.html">Provider</a> and <a href="Consumer.html">Consumer</a> components as well as a shorthand <a href="Wrapper.html">Wrapper</a> component to interact with the store.</p>
<p>It accepts a single options object, that allows the following options.</p>
<h2 id="options">Options</h2>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>actions</td>
<td><em>(Object?)</em></td>
<td>A map of <a href="../concepts/Actions.html">actions</a> where the <em>key</em> represents the action name and the <em>value</em> represents the action reducer</td>
</tr>
<tr>
<td>effects</td>
<td><em>(Object?)</em></td>
<td>A map of <a href="../concepts/Effects.html">effects</a> where the <em>key</em> represents the effect name and the <em>value</em> represents the updater</td>
</tr>
<tr>
<td>model</td>
<td><em>(any?)</em></td>
<td>The default <a href="../concepts/Model.html">model</a> reflecting the initial state shape</td>
</tr>
</tbody>
</table>
<h2 id="returns">Returns</h2>
<p><em>(Object)</em> an object containing the <a href="Provider.html">Provider</a>, <a href="Consumer.html">Consumer</a> and <a href="Wrapper.html">Wrapper</a> components. </p>
<h2 id="example">Example</h2>
<pre class="language-"><code class="lang-javascript"><span class="token keyword">import</span> <span class="token punctuation">{</span> createStore <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">&apos;alveron&apos;</span>

<span class="token keyword">const</span> model <span class="token operator">=</span> <span class="token number">0</span>
<span class="token keyword">const</span> actions <span class="token operator">=</span> <span class="token punctuation">{</span>
  increment<span class="token punctuation">:</span> state <span class="token operator">=&gt;</span> state <span class="token operator">+</span> <span class="token number">1</span><span class="token punctuation">,</span>
  decrement<span class="token punctuation">:</span> state <span class="token operator">=&gt;</span> state <span class="token operator">-</span> <span class="token number">1</span>
<span class="token punctuation">}</span>

<span class="token keyword">const</span> <span class="token punctuation">{</span> Provider<span class="token punctuation">,</span> Consumer<span class="token punctuation">,</span> Wrapper <span class="token punctuation">}</span> <span class="token operator">=</span> <span class="token function">createStore</span><span class="token punctuation">(</span><span class="token punctuation">{</span>
  actions<span class="token punctuation">,</span>
  model
<span class="token punctuation">}</span><span class="token punctuation">)</span>
</code></pre>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="../API.html" class="navigation navigation-prev " aria-label="Previous page: 4. API Reference">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
                <a href="Provider.html" class="navigation navigation-next " aria-label="Next page: 4.2. Provider">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"4.1. createStore","level":"1.5.1","depth":2,"next":{"title":"4.2. Provider","level":"1.5.2","depth":2,"path":"docs/api/Provider.md","ref":"docs/api/Provider.md","articles":[]},"previous":{"title":"4. API Reference","level":"1.5","depth":1,"path":"docs/API.md","ref":"docs/API.md","articles":[{"title":"4.1. createStore","level":"1.5.1","depth":2,"path":"docs/api/createStore.md","ref":"docs/api/createStore.md","articles":[]},{"title":"4.2. Provider","level":"1.5.2","depth":2,"path":"docs/api/Provider.md","ref":"docs/api/Provider.md","articles":[]},{"title":"4.3. Consumer","level":"1.5.3","depth":2,"path":"docs/api/Consumer.md","ref":"docs/api/Consumer.md","articles":[]},{"title":"4.4. Wrapper","level":"1.5.4","depth":2,"path":"docs/api/Wrapper.md","ref":"docs/api/Wrapper.md","articles":[]},{"title":"4.5. useAlveron","level":"1.5.5","depth":2,"path":"docs/api/useAlveron.md","ref":"docs/api/useAlveron.md","articles":[]}]},"dir":"ltr"},"config":{"plugins":["edit-link","prism","-highlight","github","anker-enable","advanced-emoji"],"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"pluginsConfig":{"prism":{},"github":{"url":"https://github.com/rofrischmann/alveron/"},"anker-enable":{},"search":{},"lunr":{"maxIndexSize":1000000},"fontsettings":{"theme":"white","family":"sans","size":2},"advanced-emoji":{"embedEmojis":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"edit-link":{"base":"https://github.com/rofrischmann/alveron/tree/master","label":"Edit This Page"},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"theme":"default","author":"Robin Frischmann <robin@rofrischmann.de>","pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"variables":{},"title":"Alveron","language":"en","gitbook":"*","description":"Official documentation for Alveron"},"file":{"path":"docs/api/createStore.md","mtime":"2019-01-23T07:37:50.760Z","type":"markdown"},"gitbook":{"version":"3.1.0","time":"2019-02-07T12:09:50.347Z"},"basePath":"../..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../../gitbook/gitbook.js"></script>
    <script src="../../gitbook/theme.js"></script>
    
        
        <script src="../../gitbook/gitbook-plugin-edit-link/plugin.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-github/plugin.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-anker-enable/anker.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

