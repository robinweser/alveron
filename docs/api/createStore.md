# createStore

The only public API as of today.<br>
It creates a new store instance that provides both [Provider](Provider.md) and [Consumer](Consumer.md) components as well as a shorthand [Wrapper](Wrapper.md) component to interact with the store.

It accepts a single options object, that allows the following options.

## Options
| Argument | Type | Description |
| -------- | ---- | ----------- |
| actions | *(Object?)* | A map of [actions](../concepts/Actions.md) where the *key* represents the action name and the *value* represents the action reducer |
| effects | *(Object?)* | A map of [effects](../concepts/Effects.md) where the *key* represents the effect name and the *value* represents the updater |
| model | *(any?)* | The default [model](../concepts/Model.md) reflecting the initial state shape |

## Returns
*(Object)* an object containing the [Provider](Provider.md), [Consumer](Consumer.md) and [Wrapper](Wrapper.md) components. 

## Example
```javascript
import { createStore } from 'react-woodworm'

const model = 0
const actions = {
  increment: state => state + 1,
  decrement: state => state - 1
}

const { Provider, Consumer, Wrapper } = createStore({
  actions,
  model
})
```