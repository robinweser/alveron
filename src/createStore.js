/* @flow */
import React, { createContext, Component } from 'react'

import validateStore from './validateStore'

type InterfaceShape = {
  Consumer: any,
  Provider: any,
}

type OptionsType = {
  model?: any,
  actions?: Object,
  effects?: Object,
}

export default function createStore(
  options: OptionsShape = {}
): InterfaceShape {
  const Store = createContext()

  const actions = options.actions || {}
  const effects = options.effects || {}

  return {
    Provider: class Provider extends Component {
      static defaultProps = {
        middleware: [],
      }

      state: {
        state: any,
        actions: Object,
        effects: Object,
      }

      constructor(props, context) {
        super(props, context)

        const initialState = props.initialState || options.model
        const resolvedActions = Object.keys(actions).reduce((map, name) => {
          map[name] = (...payload) =>
            this.setState(prevState => ({
              state: this._resolveState(
                actions[name](prevState.state, ...payload),
                {
                  action: name,
                  previousState: prevState.state,
                  payload: [...payload],
                }
              ),
            }))

          return map
        }, {})

        const resolvedEffects = Object.keys(effects).reduce((map, name) => {
          map[name] = (...payload) => effects[name](resolvedActions, ...payload)
          return map
        }, {})

        this.state = {
          state: initialState,
          actions: resolvedActions,
          effects: resolvedEffects,
        }
      }

      _resolveState = (state: any, context: Object): any => {
        return this.props.middleware.reduce(
          (finalState, middleware) => middleware(finalState, context),
          state
        )
      }

      render() {
        return (
          <Store.Provider value={this.state}>
            {this.props.children}
          </Store.Provider>
        )
      }
    },

    Consumer: ({ children }) => (
      <Store.Consumer>
        {store => validateStore(store) && children(store)}
      </Store.Consumer>
    ),
  }
}
