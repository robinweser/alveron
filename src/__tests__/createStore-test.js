import React from 'react'
import TestRenderer from 'react-test-renderer'
import { Simulate } from 'react-dom/test-utils'

import createStore from '../createStore'

describe('Creating a store', () => {
  it('should return a Consumer and a Provider', () => {
    const store = createStore()

    expect(Object.keys(store)).toEqual(['Provider', 'Consumer'])
  })
})

describe('Rendering the <Consumer>', () => {
  it('should throw without the <Provider>', () => {
    const { Provider, Consumer } = createStore()

    const render = () =>
      TestRenderer.create(<Consumer>{store => null}</Consumer>)

    expect(render).toThrow()
  })

  it('should render the initial state', () => {
    const { Provider, Consumer } = createStore({
      model: {
        foo: 1,
        bar: 'bar',
        baz: {
          foobar: 5,
        },
      },
    })

    const tree = TestRenderer.create(
      <Provider>
        <Consumer>
          {({ state }) => <div>{JSON.stringify(state, null, 2)}</div>}
        </Consumer>
      </Provider>
    )

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('should render the initial state passed to the Provider', () => {
    const { Provider, Consumer } = createStore({
      model: {
        foo: 1,
        bar: 'bar',
        baz: {
          foobar: 5,
        },
      },
    })

    const tree = TestRenderer.create(
      <Provider initialState={{ foo: 5 }}>
        <Consumer>
          {({ state }) => <div>{JSON.stringify(state, null, 2)}</div>}
        </Consumer>
      </Provider>
    )

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('should render the updated state', () => {
    const { Provider, Consumer } = createStore({
      model: {
        foo: 1,
        bar: 'bar',
        baz: {
          foobar: 5,
        },
      },
      actions: {
        update: (state, payload) => ({ ...state, foo: payload }),
      },
    })

    const tree = TestRenderer.create(
      <Provider>
        <Consumer>
          {({ state, actions }) => (
            <div>
              <div>{JSON.stringify(state, null, 2)}</div>
              <button onClick={() => actions.update(10)}>Update</button>
            </div>
          )}
        </Consumer>
      </Provider>
    )

    tree.root.findByType('button').props.onClick()

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('should support multiple arguments on actions', () => {
    const { Provider, Consumer } = createStore({
      model: 0,
      actions: {
        increment(state, stepper = 1, additionalStepper = 0) {
          return state + stepper + additionalStepper
        },
      },
    })

    const tree = TestRenderer.create(
      <Provider>
        <Consumer>
          {({ state, actions }) => (
            <div>
              <div>{JSON.stringify(state, null, 2)}</div>
              <button onClick={() => actions.increment(5, 2)}>Increment</button>
            </div>
          )}
        </Consumer>
      </Provider>
    )

    tree.root.findByType('button').props.onClick()

    expect(tree.getInstance().state.state).toBe(7)
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('should support async actions', done => {
    const { Provider, Consumer } = createStore({
      model: 10,
      actions: {
        update: state => state + 1,
        reset: () => 10,
      },
      effects: {
        resetAsync: actions => setTimeout(actions.reset, 1000),
      },
    })

    const tree = TestRenderer.create(
      <Provider>
        <Consumer>
          {({ state, actions, effects }) => (
            <div>
              <div>{state}</div>
              <button id="update" onClick={actions.update}>
                Update
              </button>
              <button id="reset" onClick={effects.resetAsync}>
                Reset After 1 second
              </button>
            </div>
          )}
        </Consumer>
      </Provider>
    )

    tree.root.findByProps({ id: 'update' }).props.onClick()
    expect(tree.toJSON()).toMatchSnapshot()

    tree.root.findByProps({ id: 'reset' }).props.onClick() // should reset to zero after 1sec

    setTimeout(() => {
      expect(tree.toJSON()).toMatchSnapshot()
      done()
    }, 1001)
  })

  it('should create a separate store for every Provider instance', () => {
    const { Provider, Consumer } = createStore({
      model: 0,
      actions: {
        increment: state => state + 1,
      },
    })

    const Counter = () => (
      <Provider>
        <Consumer>
          {({ state, actions }) => (
            <div>
              <div>Count: {state}</div>
              <button onClick={actions.increment}>+</button>
            </div>
          )}
        </Consumer>
      </Provider>
    )
    const tree = TestRenderer.create(
      <div>
        <Counter />
        <Counter />
      </div>
    )

    tree.root.findAllByType('button')[0].props.onClick()

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('should pipe state updates through middleware', () => {
    const { Provider, Consumer } = createStore({
      model: 0,
      actions: {
        increment: state => state + 1,
      },
    })

    const ghostIncrement = state => state + 1

    const tree = TestRenderer.create(
      <Provider middleware={[ghostIncrement]}>
        <Consumer>
          {({ state, actions }) => (
            <div>
              <div>{state}</div>
              <button onClick={actions.increment}>Update</button>
            </div>
          )}
        </Consumer>
      </Provider>
    )

    tree.root.findByType('button').props.onClick()
    tree.root.findByType('button').props.onClick()
    tree.root.findByType('button').props.onClick()
    tree.root.findByType('button').props.onClick()

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('should correctly pass context to middleware', () => {
    const { Provider, Consumer } = createStore({
      model: 0,
      actions: {
        increment: state => state + 1,
        decrement: state => state - 1,
        incrementBy: (state, step) => state + step,
      },
    })

    const cache = []
    const contextCache = (state, context) => {
      cache.push({ ...context, nextState: state })
      return state
    }

    const tree = TestRenderer.create(
      <Provider middleware={[contextCache]}>
        <Consumer>
          {({ state, actions }) => (
            <div>
              <div>{state}</div>
              <button id="increment" onClick={actions.increment}>
                +
              </button>
              <button id="incrementBy2" onClick={() => actions.incrementBy(2)}>
                +
              </button>
              <button id="decrement" onClick={actions.decrement}>
                -
              </button>
            </div>
          )}
        </Consumer>
      </Provider>
    )

    tree.root.findByProps({ id: 'increment' }).props.onClick()
    tree.root.findByProps({ id: 'decrement' }).props.onClick()
    tree.root.findByProps({ id: 'incrementBy2' }).props.onClick()

    expect(tree.toJSON()).toMatchSnapshot()
    expect(cache).toMatchSnapshot()
  })
})
