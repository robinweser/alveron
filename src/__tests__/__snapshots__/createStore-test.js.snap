// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the <Consumer> should correctly pass context to middleware 1`] = `
<div>
  <div>
    2
  </div>
  <button
    id="increment"
    onClick={[Function]}
  >
    +
  </button>
  <button
    id="incrementBy2"
    onClick={[Function]}
  >
    +
  </button>
  <button
    id="decrement"
    onClick={[Function]}
  >
    -
  </button>
</div>
`;

exports[`Rendering the <Consumer> should correctly pass context to middleware 2`] = `
Array [
  Object {
    "action": "increment",
    "nextState": 1,
    "payload": Array [],
    "previousState": 0,
  },
  Object {
    "action": "decrement",
    "nextState": 0,
    "payload": Array [],
    "previousState": 1,
  },
  Object {
    "action": "incrementBy",
    "nextState": 2,
    "payload": Array [
      2,
    ],
    "previousState": 0,
  },
]
`;

exports[`Rendering the <Consumer> should create a separate store for every Provider instance 1`] = `
<div>
  <div>
    <div>
      Count: 
      1
    </div>
    <button
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div>
    <div>
      Count: 
      0
    </div>
    <button
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;

exports[`Rendering the <Consumer> should pipe state updates through middleware 1`] = `
<div>
  <div>
    8
  </div>
  <button
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;

exports[`Rendering the <Consumer> should render the initial state 1`] = `
<div>
  {
  "foo": 1,
  "bar": "bar",
  "baz": {
    "foobar": 5
  }
}
</div>
`;

exports[`Rendering the <Consumer> should render the initial state passed to the Provider 1`] = `
<div>
  {
  "foo": 5
}
</div>
`;

exports[`Rendering the <Consumer> should render the updated state 1`] = `
<div>
  <div>
    {
  "foo": 10,
  "bar": "bar",
  "baz": {
    "foobar": 5
  }
}
  </div>
  <button
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;

exports[`Rendering the <Consumer> should support async actions 1`] = `
<div>
  <div>
    11
  </div>
  <button
    id="update"
    onClick={[Function]}
  >
    Update
  </button>
  <button
    id="reset"
    onClick={[Function]}
  >
    Reset After 1 second
  </button>
</div>
`;

exports[`Rendering the <Consumer> should support async actions 2`] = `
<div>
  <div>
    10
  </div>
  <button
    id="update"
    onClick={[Function]}
  >
    Update
  </button>
  <button
    id="reset"
    onClick={[Function]}
  >
    Reset After 1 second
  </button>
</div>
`;

exports[`Rendering the <Consumer> should support multiple arguments on actions 1`] = `
<div>
  <div>
    7
  </div>
  <button
    onClick={[Function]}
  >
    Increment
  </button>
</div>
`;
